feat(librarian): add repometadata package

jskeet opened the pull request:
A repometadata package is added to librarian, allowing .repo-metadata.json files to be generated.

The `DescriptionOverride` field that was previously part of `RustCrate` is moved to `Library` so it can be used by all languages (with appropriate changes to Rust tests).

jskeet commented on internal/repometadata/repometadata_test.go:
```diff
@@ -0,0 +1,318 @@
+	}
+}
+
+func TestGenerateRepoMetadata_WithOverrides(t *testing.T) {
```
I took this from the librarianx repo, where there are two separate test functions - but I can refactor it to a table-driven test if you'd prefer.

julieqiu commented on internal/config/config.go:
```diff
@@ -125,6 +125,13 @@ type Library struct {
 	// overrides Default.Transport.
 	Transport string `yaml:"transport,omitempty"`
 
+	// DescriptionOverride overrides the library description.
```
I would split this into a separate PR since it seems unrelated to repometadata

jskeet replied:
It's used by repometadata. (The original version used a Python-specific description override; this is now using a general one.)

I could:

- Leave it as-is, with that extra context
- Change this PR to not use an override at all, and then introduce a second PR which moves it from being Rust-specific to being general, and use it in repometadata at that point
- Introduce a new PR which *just* moves it from being Rest-specific to being general, and make this the "second PR".

Thoughts?

julieqiu replied:
I would do 2 or 3

julieqiu commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	// Set default release level if not specified
+	if metadata.ReleaseLevel == "" {
```
I think you should be able to use library.ReleaseLevel and that will always be accurate

jskeet commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+)
+
+// RepoMetadata represents the .repo-metadata.json file structure.
+type RepoMetadata struct {
```
I think they're already in alphabetical order, aren't they? (Wondering what I'm missing now...)

I don't know of any ordering requirements, and I really hope there aren't any.

jskeet commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+	// Build client documentation URL based on language
+	clientDocURL := buildClientDocURL(language, extractNameFromAPIID(svcCfg.GetName()))
+
+	// Create metadata
```
Done.

jskeet commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+		DefaultVersion:      defaultVersion,
+	}
+
+	// Add optional fields if available
```
Done.

jskeet commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+		return fmt.Errorf("failed to read service config: %w", err)
+	}
+
+	// Select the default version from all API paths, preferring stable versions,
```
Indeed, I'm pretty sure it will be used in Python to work out the ordering of things.

Are you suggesting "leave the code here for now, merge this PR, but then add another PR to move it" or remove everything about default versions for this first PR?

jskeet commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+		return parts[0]
+	}
+	return apiID
+}
```
Done.

jskeet commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,270 @@
+		}
+	}
+	return true
+}
```
Done (in a slightly different way) and added a test - albeit a very odd test...

julieqiu commented on internal/repometadata/repometadata.go:
```diff
@@ -0,0 +1,262 @@
+		metadata.APIDescription = strings.TrimSpace(svcCfg.GetDocumentation().GetSummary())
+	}
+
+	// Write metadata file
```
this comment doesn't seem very useful, I would just remove it

julieqiu commented on internal/repometadata/repometadata_test.go:
```diff
@@ -0,0 +1,378 @@
+
+func TestGenerateRepoMetadata_WithOverrides(t *testing.T) {
+	// Create a test service YAML
+	yamlContent := `type: google.api.Service
```
consider moving this and above into a testdata/ folder

we may want a general internal/testdata that is shared across different packages

jskeet reviewed (commented):
(Interim review comments - I still need to move some stuff to a testdata folder.)

jskeet commented:
> > This also moves the existing Rust-specific DescriptionOverride to Library (as we want it in repometadata).
> 
> There's a lot happening in this PR beyond just adding repometadata. I would describe the changes. See https://github.com/googleapis/librarian/blob/main/CONTRIBUTING.md#commit-messages

There's now *less* happening (as I've removed default versions) but I've updated the PR comment. (I think the only non-repometadata-related change is to DescriptionOverride...)

jskeet commented:
Note (as per chat) - I've only just noticed that https://github.com/googleapis/librarian/blob/main/internal/config/repometadata.go exists. (It didn't when I created the prototype that this PR is derived from.) Happy to take guidance about whether that should be deleted as part of this PR, or whether to remove the struct in my PR (but keep the code in its own package) etc.
